### PR TITLE
02 – Project Setup

### PR DESCRIPTION
**Pull request by lucasconstantino**
See [Lesson 02: Project Setup](https://github.com/strvcom/frontend-academy-2022/wiki/Lesson-02:-Project-Setup) wiki for a full lesson wrap-up.

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -1,3 +1,12 @@
 {
-  "extends": "next/core-web-vitals"
+  "extends": [
+    "next/core-web-vitals",
+    "@strv/react",
+    "@strv/typescript",
+    "@strv/typescript/react"
+  ],
```
Comment (lucasconstantino) [inline .eslintrc.json:7]: Next.js starts with a `next/core-web-vitals` ESLint config. Here, we are adding:

- [`@strv/react`](https://github.com/strvcom/code-quality-tools/blob/master/packages/eslint-config-react/index.js)
- [`@strv/typescript`](https://github.com/strvcom/code-quality-tools/blob/master/packages/eslint-config-typescript/index.js)
- [`@strv/typescript/react`](https://github.com/strvcom/code-quality-tools/blob/master/packages/eslint-config-typescript/react.js)

Follow the links above to check what this configuration looks like 😉

File: .eslintrc.json [hunk 2]
```diff
@@ -1,3 +1,12 @@
...
+
+  "parserOptions": {
+    "project": "./tsconfig.json"
+  }
```
Comment (lucasconstantino) [inline .eslintrc.json:11]: Some of the TypeScript ESLint rules we add requires that we inform where is our `.tsconfig.json` located at.

File: pages/index.tsx [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import type { NextPage } from 'next'
 import Head from 'next/head'
 import Image from 'next/image'
+
```
Comment (lucasconstantino) [inline pages/index.tsx:4]: Look! Our ESLint is already working, and enforcing some code-style. In this case, the [`import/order`](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/order.md) plugin is enforcing the [`newlines-between`](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/order.md#newlines-between-ignorealwaysalways-and-inside-groupsnever) rule, which enforces we keep empty lines in between different groups of imports: dependencies, local modules, relative modules, etc.

File: .lintstagedrc [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
+  "*.{js,ts,tsx}": [
+    "eslint --fix",
+    "prettier --write"
+  ]
+}
```
Comment (lucasconstantino) [inline .lintstagedrc:6]: This configuration is telling lint-staged to execute ESLint and Prettier on the `.js`, `.ts`, or `.tsx` files.

It also tells these tools that they can fix the issues they find, if they are able to do so (`--fix` and `--write`)

File: .husky/pre-commit [hunk 5]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+. "$(dirname "$0")/_/husky.sh"
+
+yarn lint-staged
```
Comment (lucasconstantino) [inline .husky/pre-commit:4]: This file was created automatically with the `yarn husky add .husky/pre-commit "yarn lint-staged"` command.

What is tells is that Husky should execute `yarn lint-staged` before any git commit is made (`pre-commit`)

File: package.json [hunk 6]
```diff
@@ -7,7 +7,8 @@
     "build": "next build",
     "start": "next start",
     "lint": "next lint",
-    "format": "prettier --check ."
+    "format": "prettier --check .",
+    "prepare": "husky install"
```
Comment (lucasconstantino) [inline package.json:11]: Husky is just a tool to install [Git Hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks). By adding the [`prepare`](https://docs.npmjs.com/cli/v8/using-npm/scripts#prepare-and-prepublish) script above, we ensure that anyone that installs our project will install the hooks configured under the `.husky` directory.

File: .husky/commit-msg [hunk 7]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+. "$(dirname "$0")/_/husky.sh"
+
+yarn commitlint --edit
```
Comment (lucasconstantino) [inline .husky/commit-msg:4]: This file was created automatically with the yarn husky add .husky/pre-commit "yarn commitlint --edit" command.

What is tells is that Husky should execute `yarn commitlint --edit` on top of the commit message, before creating the commit itself.

File: .gitattributes [hunk 8]
```diff
@@ -0,0 +1 @@
+yarn.lock linguist-generated=true
```
Comment (OndrejDrapalik) [inline .gitattributes:1]: @lucasconstantino This one might have slipped through my mind. Naturally, I would add yarn.lock to the .gitignore file when I would want to ignore this one. I read some articles about .gitattributes, but I'm afraid I am not yet able to see the benefit of this use case. Would you be able to clarify this one for me please? Thanks!

Comment (lucasconstantino) [reply .gitattributes:1]: Hello sir! This one I haven't really covered during our lesson, sorry, but I'll explain why. Lets go in parts:

- Should I version `yarn.lock` files?
- What is `.gitattributes`

Comment (lucasconstantino) [reply .gitattributes:1]: ## Should I version `yarn.lock` files?
The simple answer is: **yes, always version this file** (meaning, do not ignore it on `.gitignore`).

I understand this may sound weird at first, as many people – to prevent conflicts on lock files – remove it from versioning. But, this is ultimately a big misunderstanding on _why_ these files exist. At a first glance, people tend to think these files are simple "tool assets", such as `yarn-error.log`. **They are not**.

### What is `yarn.lock`
Lock files are common in many languages and frameworks. The correct name for them is **Dependency Lock File**. These files are meant hold information on **precise versions** of installed dependencies, and **integrity hashes** to ensure whenever we reinstall the project, we download the _exact same dependency codebase_ as defined by this file.

Why is this even a problem? Well, if you look into your `package.json`, you will notice our dependencies are described with a name, and version definition. The version definition, however, is "loose". For example: `^1.0.0` means that you are requesting Yarn/NPM to install this package in a version that's bigger than `1.0.0`, but flexible in terms of minor releases and patches. For all Yarn/NPM know, version `1.89.412` would be completely fine for you. And well, it should: if only package maintainers where trustworthy and followed [semantic versioning](https://docs.npmjs.com/about-semantic-versioning#using-semantic-versioning-to-specify-update-types-your-package-can-accept) correctly.

The problem is that _you can't trust your dependencies_. And I'm not judging the maintainers of open-source projects, praise them please, but _they are human and fail_.

A lock file saves the exact version _you yourself installed_ at your development machine. So, in theory, you know this version works perfectly fine within your project. You don't want your CI/CD to install a new version of this dependency when you deploy, because you can't be sure it will work as well as the version you had locally.

Besides this, lock files also maintain hashes that are a representation of the codebase. This means if some malicious person somehow replaces the dependency codebase on the source you are downloading from, the hashes will be different, and the install will fail.

### When **not** to version `yarn.lock`
Lock files are for the _root_ application. Meaning, you shouldn't have a lock file versioned if you are building and publishing a library to be installed by others. In fact, these lock files would just be ignored, and besides that, in case you are building open-source chances are you actually want to be constantly updating your dependencies to the latests, in theory safest versions.

Comment (lucasconstantino) [reply .gitattributes:1]: ## What is [`.gitattributes`](https://git-scm.com/docs/gitattributes)

This file is meant for giving a path tracked by git some information on what it is and how it should be versioned. It's useful for a multitude of needs:

- Better version of complex files (designs, binaries, etc)
- [Encrypting](https://github.com/AGWA/git-crypt) versioned files
- Tagging generated files
- etc

In this case, I'm telling git itself that the `yarn.lock` file is _generated_, meaning, it is created and maintained by some tool, not humans. There aren't many benefits of this in your development cycle, but guess what, [GitHub itself uses this information](https://docs.github.com/en/repositories/working-with-files/managing-files/customizing-how-changed-files-appear-on-github) to determine if it should show diffs for this file on it's own UI.

Basically, I'm telling GitHub to fold the diffs on this files, cause we don't wanna bother with this when reviewing PRs.

Comment (danbalarin) [reply .gitattributes:1]: Regarding the `.gitattributes`, you should always consider whether you actually want to hide this from PR. For most of the applications you don't really go through the changes there, but from my experience in banks it's common practice to go dependency to dependency in lock file and check why the change occurred and whether it _"legit"_.
Ofc it's better to leave this concern for some tool like [Snyk](https://snyk.io/).
